examples/xds-management-server/pkg/providers/file.go

```diff
@@ -0,0 +1,309 @@
+package providers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	envoylistener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	gogojsonpb "github.com/gogo/protobuf/jsonpb"
+	prototypes "github.com/gogo/protobuf/types"
+	"google.golang.org/protobuf/encoding/protojson"
+	"quilkin.dev/xds-management-server/pkg/cluster"
+	"quilkin.dev/xds-management-server/pkg/filterchain"
+	"sigs.k8s.io/yaml"
+
+	"github.com/cenkalti/backoff"
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+)
+
+// FilterConfig represents a filter's config.
+type FilterConfig interface{}
+type resources struct {
+	Clusters    []cluster.Cluster
+	FilterChain []FilterConfig
+}
+
+// FileProvider watches a file on disk for resources.
+type FileProvider struct {
+	configFilePath string
+	proxyIDCh      <-chan string
+}
+
+// NewFileProvider creates a new FileProvider.
+func NewFileProvider(configFilePath string, proxyIDCh chan string) *FileProvider {
+	return &FileProvider{
+		configFilePath: configFilePath,
+		proxyIDCh:      proxyIDCh,
+	}
+}
+
+// Run runs the FileProvider.
+func (p *FileProvider) Run(ctx context.Context, logger *log.Logger) (
+	<-chan []cluster.Cluster, <-chan filterchain.ProxyFilterChain, <-chan error) {
+	clusterCh := make(chan []cluster.Cluster, 10)
+	filterChainCh := make(chan filterchain.ProxyFilterChain, 10)
+	resourcesCh := make(chan resources)
+
+	errorCh := make(chan error)
+
+	go runUpdater(
```

Comment:
A thought: Both `runUpdater` and `runConfigFileProvier` could be methods on `p` - that would save you passing in both the `logger` and `proxyIDCh` since they would be accessible.